\documentclass[12pt]{article}
\usepackage[margin=1in]{geometry}
\usepackage[all]{xy}

\usepackage{amsmath,amsthm,amssymb,color,latexsym,soul}
\usepackage{geometry}        
\geometry{letterpaper}    
\usepackage{graphicx}
\usepackage{enumitem}
\usepackage{listings}
\usepackage{xcolor}
\usepackage{bm}
\usepackage{hyperref}
\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    filecolor=magenta,      
    urlcolor=cyan,
    citecolor=blue,
}

\definecolor{codegreen}{rgb}{0,0.6,0}
\definecolor{codegray}{rgb}{0.5,0.5,0.5}
\definecolor{codepurple}{rgb}{0.58,0,0.82}
\definecolor{backcolour}{rgb}{0.95,0.95,0.92}

\lstdefinestyle{mystyle}{ % taken from: https://www.overleaf.com/learn/latex/Code_listing
    backgroundcolor=\color{backcolour},   
    commentstyle=\color{codegreen},
    keywordstyle=\color{magenta},
    numberstyle=\tiny\color{codegray},
    stringstyle=\color{codepurple},
    basicstyle=\ttfamily\footnotesize,
    % breakatwhitespace=false,         
    breaklines=true,                 
    captionpos=b,                    
    keepspaces=true,                 
    % numbers=left,                    
    numbersep=5pt,                  
    % showspaces=false,                
    showstringspaces=false,
    % showtabs=false,                  
    % tabsize=2
}

\lstset{style=mystyle}

\newtheorem{task}{Task}
\newenvironment{solution}[1][\it{Solution}]{\textbf{#1. } }{$\square$}
\newtheorem{subtask}{\; \; \it{Part}}


\begin{document}
\noindent Asaad Mohammedsaleh \hfill CS249 Assignment 2\\
KAUST Spring 2025 \hfill Genome Assembly and Evaluation


\hrulefill

\section{Introduction}

In this assignment, I had the opportunity to work with assembling and evaluating genome assemblies. 

In the first part of the assignment, I implemented the De Bruijn Graph (DBG) and Overlap-Layout-Consensus (OLC) algorithms.
The assignment sheet provided a toy dataset to test the algorithms and a larger more realistic dataset of Middle East respiratory
syndrome-related coronavirus (MERS-CoV) genome. I evaluated the performance of these algorithms on the datasets and compared them against each other and against a more established tool, SPAdes.

In the second part of the assignment, I worked with a Scincus mitranus lizard genome and evaluated the performance of the assembly using the tools described in the assignment sheet. 

The implementation of the algorithms and the evaluation of the performance are described in the following sections. The code is available at \url{https://github.com/Asaad47/BioAlgos-Assignment2}.

\section{Task 1.1 De Bruijn Graph (DBG) Assembly}

Requirements of this algorithm are:
\begin{itemize}
    \item Takes FASTQ files as input
    \item Constructs a de Bruijn graph from k-mers (with user-defined k)
    \item Identifies contigs by finding Eulerian paths
    \item Outputs contigs as a FASTA file
\end{itemize}

The implementation of the algorithm can be found in \texttt{src/dbg.go} file. The main function to handle the algorithm's components is \texttt{DBGAssembler}.
Then, there are two helper functions: \texttt{constructDeBruijnGraph} and \texttt{walkGraph}.

\texttt{constructDeBruijnGraph} constructs the de Bruijn graph from the reads. It first creates a map of k-mers to their nodes. Then, it adds the edges to the graph. The graph is represented as a map of k-mers to their nodes, where each node is a struct with the k-mer string, its outgoing edges (a map of other k-mers to the number of times they appear in the reads), and its incoming edges (a map of other k-mers to the number of times they appear in the reads).

\texttt{walkGraph} looks at nodes with no incoming edges and picks the longest outgoing edge, which would be a greedy approach on the most occurence of the k-mer. It then follows the edge to the next node and repeats the process until it reaches a node with no outgoing edges. It then adds the collected contig from this path to the list of contigs. It repeats this process for all nodes with no incoming edges, so essentially the output has the same number of contigs as the number of nodes with no incoming edges.

\texttt{DBGAssembler} then writes the contigs to a FASTA file.

\section{Task 1.2 Overlap-Layout-Consensus (OLC) Assembly}

Requirements of this algorithm are:
\begin{itemize}
    \item Takes FASTQ files as input
    \item Computes all-vs-all read overlaps and constructs an overlap graph,
    using the minimum overlap length n as a parameter
    \item Identifies non-branching paths in the graph
    \item Generates a layout of reads
    \item Computes a consensus sequence for each contig
    \item Outputs contigs as a FASTA file
\end{itemize}

The implementation of the algorithm can be found in \texttt{src/olc.go} file. The main function to handle the algorithm's components is \texttt{OLCAssembler}. Then, there are three helper functions: \texttt{overlap}, \texttt{layout}, and \texttt{consensus}.

\texttt{overlap} computes all-vs-all read overlaps and constructs an overlap graph, using the minimum overlap length \texttt{min\_overlap} as a parameter. It returns a map of reads to their nodes, where each node is a struct with the read string, its outgoing edges (a map of other reads to the largest overlap size), and its incoming edges (a map of other reads to the largest overlap size).
The implementation is a naive approach that iterates over all pairs of reads and checks for overlaps, which worked sufficiently for the assignment's datasets.

\texttt{layout} reduces the size of the overlap graph by removing 1-hop and 2-hop inferrible edges as described in the lecture notes of Johns Hopkins slides \cite{olc_lecture_notes}. During the process, it updates the overlap graph by combining reads that have single in-edges and single out-edges. It then outputs the list of reads of the updated overlap graph as contigs.

\texttt{consensus} is a greedy algorithm that starts from a read with no incoming edges and follows the longest outgoing edge until it reaches a read with no outgoing edges. It then adds the collected contig from this path to the list of contigs. It repeats this process for all reads with no incoming edges, so essentially the output has the same number of contigs as the number of reads with no incoming edges. This is not an optimal solution, but it happened to be a good approximation for the assignment's datasets.



\section{Task 1.3 Applications of assembly algorithms}

\subsection{Task 1.3.1}

\begin{figure}[h!]
    \centering
    \includegraphics[width=0.5\textwidth]{../toy_dataset/reads_b_k_40.png}
    \caption{Bandage visualization of DBG for $k=40$}
    \label{fig:dbg_k_40}
\end{figure} 

Figure \ref{fig:dbg_k_40} shows the DBG for $k=40$ on reads\_b.fastq visualized using Bandage. A bubble can be seen in the middle of the graph, which suggests that the assembly is not optimal to reconstruct the original genome.
This hints toward using a larger $k$ value to get a better assembly, where in the next section, we will see that using $k=45$ gives a better assembly.

\subsection{Task 1.3.2}

\begin{figure}[h!]
    \centering
    \includegraphics[width=0.5\textwidth]{../toy_dataset/r-k-35.png}
    \caption{Bandage visualization of DBG contigs for $k=35$ on reads\_r.fastq}
    \label{fig:dbg_k_35}
\end{figure} 

\begin{figure}[h!]
    \centering
    \includegraphics[width=0.5\textwidth]{../toy_dataset/r-k-45.png}
    \caption{Bandage visualization of DBG contigs for $k=45$ on reads\_r.fastq}
    \label{fig:dbg_k_45}
\end{figure} 

\begin{table}[h!]
\begin{center}
\begin{tabular}{ |c|c|c| }
    \hline
    Metric & k=35 & k=45 \\
    \hline
    sequence length & 1020 & 1040 \\
    number of contigs & 1 & 1 \\
    GC content (\%) & 51.37 & 51.25 \\
    genome fraction (\%) & 100.000 & 100.000 \\
    duplication ratio & 0.981 & 1.000 \\
    largest contig & 1020 & 1040 \\
    N50 & 1020 & 1040 \\
    N90 & 1020 & 1040 \\
    L50 & 1 & 1 \\
    misassemblies & 0 & 0 \\
    mismatches per 100 kbp & 0.00 & 0.00 \\
    indels per 100 kbp & 98.04 & 0.00 \\
    \hline
\end{tabular}
\caption{QUAST evaluation for $k=35$ and $k=45$ against reference\_r.fasta}
\label{tab:quast_k35_k45}
\end{center}
\end{table}

Figures \ref{fig:dbg_k_35} and \ref{fig:dbg_k_45} show the DBG for $k=35$ and $k=45$ on reads\_r.fastq visualized using Bandage. The DBG for $k=35$ has a bubble in the middle, which suggests that the assembly is not optimal to reconstruct the original genome. The DBG for $k=45$ has a more compact structure of a single contig, which suggests that the assembly is more optimal to reconstruct the original genome.
Table \ref{tab:quast_k35_k45} shows the QUAST evaluation for $k=35$ and $k=45$ against reference\_r.fasta. The DBG for $k=45$ has a perfect match to the reference genome, which can be seen in the genome fraction and duplication ratio. The DBG for $k=35$ has a lower duplication ratio and a sequence length but a perfect genome fraction, which suggest that the assembly is only missing a few nucleotides to completely match the reference genome.
The missed nucleotides are likely due to the bubble in the middle of the DBG for $k=35$.


\subsection{Task 1.3.3}

\subsubsection{ONT reads}
For DBG, I used $k = 40$. For OLC, I used $m = \text{min\_overlap} = 40$. These values had better assemblies than other values I tried (e.g. $k = 35,45$, etc. and $m = 30,50$, etc.).

Table \ref{tab:quast_mers} shows the QUAST evaluation for DBG and OLC on MERS virus data. The DBG has a better assembly than the OLC, which can be seen in the genome fraction and duplication ratio. The OLC has a higher duplication ratio and a lower genome fraction, which suggests that the implemented OLC algorithm is not as good at reconstructing the original genome.
Between the error-free and errored reads, both algorithms achieve obvious better performance in the error-free case bu having lower number of contigs and lower number of duplication ratios while having a high genome fraction. The OLC implementation is quite bad at reconstructing the original genome in both cases. In the errored reads case, OLC only reduced the number of contigs by 3. 
This is heavily influenced by the fact that the OLC \texttt{consensus} function is not optimized to solve the problem of reconstructing the original genome when having errors and small number of reads. I tried to implement a better \texttt{consensus} function, but it did not improve the performance of the OLC algorithm.


\begin{table}[h!]
\begin{center}
\begin{tabular}{ |c|c|c||c|c| }
    \hline
    Metric               & DBG (no-errors) & OLC (no-errors) & DBG (errors) & OLC (errors) \\
    \hline
    sequence length      & 29748  & 121258  & 76608 & 1505783 \\
    number of contigs    & 1      & 4       & 11 & 163 \\
    GC content (\%)      & 41.27  & 41.22    & 41.21 & 40.84 \\
    genome fraction (\%) & 98.765 & 96.670    & 98.738 & 98.655 \\
    duplication ratio    & 1.000  & 4.164    & 1.973 & 37.597 \\
    largest contig       & 29748  & 80399    & 17192 & 20439 \\
    N50                  & 29748  & 80399    & 12088 & 9201 \\
    N90                  & 29748  & 10145    & 1713 & 7834 \\
    L50                  & 1      & 1       & 3 & 73 \\
    misassemblies        & 0      & 29        & 0 & 0 \\
    mismatches per 100 kbp & 0.00 & 0.00     & 78.39 & 493.40 \\
    indels per 100 kbp   & 3.36   & 0.00     & 184.04 & 1463.10  \\
    \hline
\end{tabular}
\end{center}
\caption{QUAST evaluation for DBG and OLC on MERS virus data with ONT reads}
\label{tab:quast_mers}
\end{table}

In finding an optimal $m$ value for OLC, I tried different values of $m$ and found that $m = 3$ and $m = 2$ gave interesting results. Table \ref{tab:quast_mers_olc_m3_m2} shows the QUAST evaluation for OLC with $m = 3$ and $m = 2$ on MERS virus data.

They heavily reduce the number of contigs found while having high genome fractions. However, they both have high duplication ratios and high indels per 100 kbp, which suggests that they are not as good at reconstructing the original genome as DBG implementation.

Further work could be done to improve the OLC algorithm to reconstruct the original genome when having errors and small number of reads.

\begin{table}[h!]
\begin{center}
    \begin{tabular}{ |c|c|c| }
        \hline
        Metric               & m = 3 & m = 2 \\
        \hline
        sequence length      & 999925  & 537140   \\
        number of contigs    & 57      & 13        \\
        GC content (\%)      & 40.83  & 40.90    \\
        genome fraction (\%) & 97.600 & 94.767    \\
        duplication ratio    & 24.110  & 14.208    \\
        largest contig       & 73843  & 62932    \\
        N50                  & 19867  & 54062    \\
        N90                  & 8207  & 25366    \\
        L50                  & 13      & 5       \\
        misassemblies        & 36      & 34        \\
        mismatches per 100 kbp & 437.82 & 463.10     \\
        indels per 100 kbp   & 1359.61   & 1411.50     \\
        \hline
    \end{tabular}
    \end{center}
\caption{QUAST evaluation for OLC with $m = 3$ and $m = 2$ on MERS virus data}
\label{tab:quast_mers_olc_m3_m2}
\end{table}

\subsubsection{HiSeq reads}
Similar to the ONT reads, I used $k = 40$ for DBG and $m = \text{min\_overlap} = 40$ for OLC. Also, these values had better assemblies than other values I tried (e.g. $k = 35,45$, etc. and $m = 30,50$, etc.).

Table \ref{tab:quast_mers_dbg_olc} shows the QUAST evaluation for DBG and OLC on MERS virus data. In this case, both DBG and OLC have similar performance on the no-error case. However, OLC implementation fails horribly in the errored reads case.
This is also heavily due to the fact that the OLC \texttt{consensus} function is not optimized to solve the problem of reconstructing the original genome when having errors. I tried to implement a better \texttt{consensus} function, but it did not improve the performance of the OLC algorithm.

On the other hand, DBG implementation is able to reconstruct the original genome in the errored case relatively well with 94.425\% genome fraction and 1.014 duplication ratio. Also, relatively low number of contigs and indels per 100 kbp.

\begin{table}[h!]
\begin{center}
\begin{tabular}{ |c|c|c||c|c| }
    \hline
    Metric & DBG (no-errors) & OLC (no-errors) & DBG (errors) & OLC (errors) \\
    \hline
    sequence length        & 29553  & 29553  & 28835 & 18535 \\
    number of contigs      & 3      & 3      & 11 & 24 \\
    GC content (\%)        & 41.26  & 41.27  & 41.20 & 40.71 \\
    genome fraction (\%)   & 97.882 & 97.885 & 94.425 & 29.022 \\
    duplication ratio      & 1.002  & 1.002  & 1.014 & 1.157 \\
    largest contig         & 12290  & 12290  & 5155 & 2467 \\
    N50                    & 8725   & 8725   & 4682 & 734 \\
    N90                    & 8538   & 8538   & 1684 & 562 \\
    L50                    & 2      & 2      & 3 & 10 \\
    misassemblies          & 0      & 0 & 0  & 0 \\
    mismatches per 100 kbp & 0.00   & 0.00   & 38.16 & 692.11 \\
    indels per 100 kbp     & 10.15  & 0.00   & 38.16 & 266.96 \\
    \hline
\end{tabular}
\end{center}
\caption{QUAST evaluation for DBG and OLC on MERS virus data with HiSeq reads}
\label{tab:quast_mers_dbg_olc}
\end{table}

\subsection{Task 1.3.4}

Table \ref{tab:quast_mers_spades} shows the QUAST evaluation for SPAdes on MERS virus data. SPAdes is able to reconstruct the original genomes with high genome fractions and using only one contig assembly.

In the case of ONT reads, SPAdes and DBG have similar performance with similar genome fractions in the error-free and errored cases, but SPAdes has consistently lower duplication ratios and indels per 100 kbp, which can be understood as SPAdes only using one contig assembly.

In the case of HiSeq reads, SPAdes, DBG, and OLC have similar performance with similar genome fractions and GC content in the error-free case. However, in the errored case, SPAdes achieves better performance than DBG and OLC across all metrics.
This is expected as SPAdes has been developed and designed more rigorously while my implementations are simple and naive.

\begin{table}[h!]
\begin{center}
\begin{tabular}{ |c|c|c|c|c| }
    \hline
    Metric & ONT (no-errors) & ONT (errors) & HiSeq (no-errors) & HiSeq (errors) \\
    \hline
    sequence length & 29748 & 29751 & 29482 & 29482 \\
    number of contigs & 1 & 1 & 1 & 1 \\
    GC content (\%) & 41.27 & 41.25 & 41.26 & 41.26  \\
    genome fraction (\%) & 98.768 & 98.738 & 97.885  & 97.885 \\
    duplication ratio & 1.000 & 1.000 & 1.000 & 1.00 \\
    largest contig & 29748 & 29751 & 29482 & 29482 \\
    N50 & 29748 & 29751 & 29482 & 29482 \\
    N90 & 29748 & 29751 & 29482 & 29482 \\
    L50 & 1 & 1 & 1 & 1 \\
    misassemblies & 0 & 0 & 0 & 0 \\
    mismatches per 100 kbp & 0.00 & 0.00 & 0.00 & 0.00 \\
    indels per 100 kbp & 0.00 & 0.00 & 0.00 & 0.00 \\
    \hline
\end{tabular}
\end{center}
\caption{QUAST evaluation for SPAdes on MERS virus data}
\label{tab:quast_mers_spades}
\end{table}

\section{Task 2.1 Lizard Genome Assembly}

The commands used to assemble the lizard genome are described in the following bash commands. The assembly took around 9 hours and 16 minutes.

I used the allocated resources for the class in Ibex cluster. The resources used were (part of the head of my slurm script):

\begin{lstlisting}[language=bash]
#SBATCH --cpus-per-task=64
#SBATCH --mem=256G
#SBATCH --account=cs249
\end{lstlisting}

The assembly was done using the following commands:

\begin{lstlisting}[language=bash]
# 1- Setup
module load gcc/12.2.0 hifiasm/0.19 minimap2 samtools seqtk
mkdir -p asm_scinmit/{assembly,data,eval,tmp}
cd asm_scinmit

# 2- Fetch data
export RAW=/ibex/reference/course/cs249/lizard/input
# Create symbolic links inside project tree
ln -s $RAW/pacbio/*.fastq.gz data/pacbio/
ln -s $RAW/hic/*.fastq.gz data/hic/
ln -s $RAW/ont/*.fastq.gz data/ont/

# 3- Assemble primary contigs (took around 9 hours and 16 minutes)
# Real time: 33467.576 sec; CPU: 1523075.897 sec; Peak RSS: 170.915 GB
hifiasm -o assembly/scinmit_hicul \
        -t64 --primary -D10 \
        --h1 data/hic/lizard_hic_R1.fastq.gz \
        --h2 data/hic/lizard_hic_R2.fastq.gz \
        --ul data/ont/lizard_ont.fastq.gz \
        data/pacbio/lizard_liver_seq.fastq.gz \
        data/pacbio/lizard_liver_rev.fastq.gz

# 4- Convert GFA to FASTA
awk '$1=="S"{print ">"$2"\n"$3}' assembly/scinmit_hicul.hic.p_ctg.gfa > assembly/scinmit_hicul.hic.p_ctg.fa
\end{lstlisting}

\texttt{hifiasm} \cite{hifiasm} took the \_seq and \_rev liver files as input reads, in addition to the Hi-C and ultra-long ONT reads. 

\section{Task 2.2 Assembly Evaluation}

For the evaluation tools, I used QUAST \cite{quast}, BUSCO \cite{busco}, Merqury \cite{merqury}, and Inspector \cite{inspector}. 
Commands used to run the tools are described in the code repository \url{https://github.com/Asaad47/BioAlgos-Assignment2#task-22-evaluation-scripts}.

\subsection{QUAST evaluation}

Found in \texttt{eval/quast/report.pdf}. 

Table \ref{tab:quast_scinmit_hicul} shows the QUAST evaluation for scinmit\_hicul.hic.p\_ctg.fa.

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & scinmit\_hicul.hic.p\_ctg.fa \\
        \hline
        sequence length & 1807572148 \\
        number of contigs & 94 \\
        GC content (\%) & 45.48 \\
        largest contig & 341760416 \\
        N50 & 138345715 \\
        N90 & 40553331 \\
        L50 & 4 \\
        \hline
    \end{tabular}
    \caption{\textbf{QUAST} evaluation for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:quast_scinmit_hicul}
    \end{center}
    \end{table}

\subsection{BUSCO evaluation}

Found in \texttt{eval/busco/short\_summary.specific.sauropsida\_odb10.busco.json}. Reported against \texttt{sauropsida\_odb10} lineage.

Table \ref{tab:busco_scinmit_hicul} shows the BUSCO evaluation for scinmit\_hicul.hic.p\_ctg.fa.

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & Value \\
        \hline
        Complete percentage & 98.2 \\
        Complete BUSCOs & 7349 \\
        Single copy percentage & 97.4 \\
        Single copy BUSCOs & 7285 \\
        Multi copy percentage & 0.9 \\
        Multi copy BUSCOs & 64 \\
        Fragmented percentage & 0.2 \\
        Fragmented BUSCOs & 17 \\
        Missing percentage & 1.5 \\
        Missing BUSCOs & 114 \\
        n\_markers & 7480 \\
        avg\_identity & 0.81 \\
        domain & eukaryota \\
        internal\_stop\_codon\_count & 178 \\
        internal\_stop\_codon\_percent & 2.4 \\
        \hline
    \end{tabular}
    \caption{\textbf{BUSCO} evaluation for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:busco_scinmit_hicul}
    \end{center}
    \end{table}

\subsection{Merqury evaluation}

Found in \texttt{eval/merqury.qv} and \texttt{eval/merqury.scinmit\_hicul.hic.p\_ctg.spectra-cn.hist}. 

Table \ref{tab:merqury_scinmit_hicul} shows the Merqury evaluation for scinmit\_hicul.hic.p\_ctg.fa.
Figure \ref{fig:k_mer_spectrum} and \ref{fig:k_mer_spectrum_zoom} show the K-mer spectrum for scinmit\_hicul.hic.p\_ctg.fa.
The graphs were generated using the \texttt{plot\_kmer\_spectra.py} script.

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & Value \\
        \hline
        Number of false k-mers & 10415 \\
        Total length & 1807575410 \\
        QV & 65.6166 \\
        Estimated error rate & 2.74375e-07 \\
        \hline
    \end{tabular}
    \caption{\textbf{Merqury} evaluation for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:merqury_scinmit_hicul}
    \end{center}
    \end{table}

\begin{figure}[h!]
    \centering
    \begin{minipage}[t]{0.48\textwidth}
        \centering
        \includegraphics[width=\textwidth]{k_mer_spectrum.png}
        \caption{\textbf{Merqury} K-mer spectrum for \\ scinmit\_hicul.hic.p\_ctg.fa}
        \label{fig:k_mer_spectrum}
    \end{minipage}
    \hfill
    \begin{minipage}[t]{0.48\textwidth}
        \centering
        \includegraphics[width=\textwidth]{k_mer_spectrum_zoomed.png}
        \caption{Zoomed K-mer spectrum for scinmit\_hicul.hic.p\_ctg.fa}
        \label{fig:k_mer_spectrum_zoom}
    \end{minipage}
\end{figure}

\subsection{Inspector evaluation}

Found in \texttt{eval/inspector/summary\_statistics}.

Tables \ref{tab:inspector_scinmit_hicul_alignment}, \ref{tab:inspector_scinmit_hicul_structural_errors}, \ref{tab:inspector_scinmit_hicul_small_scale_errors}, and \ref{tab:inspector_scinmit_hicul_qv} show the Inspector evaluation for scinmit\_hicul.hic.p\_ctg.fa.

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & Value \\
        \hline
        Mapping rate /\% & 99.88 \\
        Split-read rate /\% & 3.76 \\
        Depth & 68.5504 \\
        Mapping rate in large contigs /\% & 99.18 \\
        Split-read rate in large contigs /\% & 3.79 \\
        Depth in large conigs & 68.4291 \\
        \hline
    \end{tabular}

    \caption{\textbf{Inspector} statistics of read to contig alignment for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:inspector_scinmit_hicul_alignment}
    \end{center}
    \end{table}

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & Value \\
        \hline
        Structural error & 126 \\
        Expansion & 60 \\
        Collapse & 42 \\
        Haplotype switch & 15 \\
        Inversion & 9 \\
        \hline
    \end{tabular}
    \caption{\textbf{Inspector} statistics of structural errors for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:inspector_scinmit_hicul_structural_errors}
    \end{center}
    \end{table}

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & Value \\
        \hline
        Small-scale assembly error /per Mbp & 5086.222664590693 \\
        Total small-scale assembly error & 9193668 \\
        Base substitution & 8131903 \\
        Small-scale expansion & 590427 \\
        Small-scale collapse & 471338 \\
        \hline
    \end{tabular}
    \caption{\textbf{Inspector} statistics of small-scale assembly errors for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:inspector_scinmit_hicul_small_scale_errors}
    \end{center}
    \end{table}

\begin{table}[h!]
    \begin{center}
    \begin{tabular}{ |c|c| }
        \hline
        Metric & Value \\
        \hline
        QV & 22.897974097278006 \\
        \hline
    \end{tabular}
    \caption{\textbf{Inspector} QV for scinmit\_hicul.hic.p\_ctg.fa}
    \label{tab:inspector_scinmit_hicul_qv}
    \end{center}
    \end{table}

\subsection{Analysis}

The assembly of the Scincus mitranus lizard genome shows promising results across multiple evaluation metrics:

\subsubsection{Assembly Structure and Completeness}
\begin{itemize}
    \item The assembly produced 94 contigs with a total length of 1.81 Gbp, which is reasonable for a reptile genome (compared to 3.2 Gbp for the human genome and similar to sauropsida's genome).
    \item The N50 of 138.3 Mbp and N90 of 40.6 Mbp indicate good contiguity, with the largest contig being 341.8 Mbp.
    \item The GC content of 45.48\% is within the expected range for vertebrate genomes.
\end{itemize}

\subsubsection{Gene Content and Completeness}
\begin{itemize}
    \item BUSCO analysis against the sauropsida lineage shows excellent completeness:
    \begin{itemize}
        \item 98.2\% complete BUSCOs (7349 genes)
        \item 97.4\% single-copy genes (7285)
        \item Only 1.5\% missing genes (114)
    \end{itemize}
    \item The high single-copy percentage (97.4\%) suggests good haplotype resolution.
    \item The presence of 64 multi-copy genes (0.9\%) might indicate some duplication or assembly artifacts.
\end{itemize}

\subsubsection{Assembly Quality}
\begin{itemize}
    \item Merqury analysis shows:
    \begin{itemize}
        \item High QV of 65.62, indicating good base-level accuracy
        \item Low estimated error rate of 2.74e-07
        \item The k-mer spectrum shows a clear peak at ~380 coverage, which supports the diploid nature of the genome since the input contig is only the primary contig.
    \end{itemize}
    \item Inspector analysis reveals:
    \begin{itemize}
        \item Excellent mapping rate (99.88\%) and depth (68.55x)
        \item 126 structural errors, with expansions (60) being more common than collapses (42)
        \item Small-scale errors are present but at a reasonable rate (5086 per Mbp)
        \item The QV of 22.90 from Inspector is lower than Merqury's estimate, suggesting some disagreement between the tools
    \end{itemize}
\end{itemize}

\subsubsection{Overall Assessment}
The assembly appears to be of high quality, with:
\begin{itemize}
    \item Excellent gene completeness (98.2\% BUSCO)
    \item Good contiguity (N50 $>$ 138 Mbp)
    \item High base-level accuracy (QV $>$ 65)
    \item Good haplotype resolution (97.4\% single-copy genes)
\end{itemize}

Areas for potential improvement include:
\begin{itemize}
    \item Reducing the number of structural errors (126 reported by Inspector)
    \item Resolving the discrepancy between Merqury and Inspector QV estimates
    \item Further investigation of the 64 multi-copy genes
\end{itemize}

Overall, this assembly provides a solid foundation for downstream analyses of the Scincus mitranus genome.

\section{Assembled Genome Links}

The assembled genome is available at this directory:

\texttt{/ibex/user/mohaah0a/cs249/BioAlgos-Assignment2/asm\_scinmit/assembly}

With important files:

\begin{itemize}
    \item \texttt{scinmit\_hicul.hic.p\_ctg.fa} - the primary contig assembly.
    \item \texttt{scinmit\_hicul.hic.a\_ctg.fa}
    \item \texttt{scinmit\_hicul.hic.hap1.p\_ctg.fa}
    \item \texttt{scinmit\_hicul.hic.hap2.p\_ctg.fa}
\end{itemize}

\begin{thebibliography}{9}
\bibitem{olc_lecture_notes}
Langmead, B. (2024). Overlap-Layout-Consensus Assembly. Johns Hopkins University. 
\url{https://www.cs.jhu.edu/~langmea/resources/lecture_notes/assembly_olc.pdf}

\bibitem{hifiasm}
Cheng, H., Concepcion, G.T., Feng, X., Zhang, H., Li H. (2021) Haplotype-resolved de novo assembly using phased assembly graphs with hifiasm. 
Nat Methods, 18:170-175. \url{https://doi.org/10.1038/s41592-020-01056-5}

\bibitem{quast}
Alla Mikheenko, Vladislav Saveliev, Pascal Hirsch, Alexey Gurevich,
WebQUAST: online evaluation of genome assemblies,
Nucleic Acids Research (2023) 51 (W1): W601-W606. \url{https://doi.org/10.1093/nar/gkad406}

\bibitem{busco}
Mosè Manni, Matthew R Berkeley, Mathieu Seppey, Felipe A Simão, Evgeny M Zdobnov, 
BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. 
Molecular Biology and Evolution, Volume 38, Issue 10, October 2021, Pages 4647-4654. \url{https://doi.org/10.1093/molbev/msab181}

\bibitem{merqury}
Rhie, A., Walenz, B.P., Koren, S. et al. Merqury: reference-free quality, completeness, and phasing assessment for genome assemblies.
 Genome Biol 21, 245 (2020). \url{https://doi.org/10.1186/s13059-020-02134-9}

\bibitem{inspector}
Chen, Y., Zhang, Y., Wang, A.Y. et al. Accurate long-read de novo assembly evaluation with Inspector.
Genome Biol 22, 312 (2021). \url{https://doi.org/10.1186/s13059-021-02527-4}

\end{thebibliography}

\end{document}
